### ISSUE TITLE
Segmentation fault ending training run

### ISSUE THREAD

**Issue by jamesgwen** [MIC-DKFZ/nnUNet#2257]
Hello!

I am currently using the MSD pancreas dataset to get to know how to use nnUNet. When I attempt to start a training run, the command doesn't error out but stops. The last line the console prints out is "Segmentation fault". However, I could not find where this print statement is coming from in the repo. Any advice on this "error" is appreciated. 

I am using the following command to start training: 

```
nnUNetv2_train 007  3d_fullres 0 --npz
```
and get the following output: 
```############################
INFO: You are using the old nnU-Net default plans. We have updated our recommendations. Please consider using those instead! Read more here: https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/resenc_presets.md
############################

Using device: cuda:0

#######################################################################
Please cite the following paper when using nnU-Net:
Isensee, F., Jaeger, P. F., Kohl, S. A., Petersen, J., & Maier-Hein, K. H. (2021). nnU-Net: a self-configuring method for deep learning-based biomedical image segmentation. Nature methods, 18(2), 203-211.
#######################################################################

2024-06-04 00:42:28.511859: do_dummy_2d_data_aug: True
2024-06-04 00:42:28.513456: Using splits from existing split file: /home/jupyter/decathalon_processed/nnUNet_preprocessed/Dataset007_Pancreas/splits_final.json
2024-06-04 00:42:28.513786: The split file contains 5 splits.
2024-06-04 00:42:28.513843: Desired fold for training: 0
2024-06-04 00:42:28.513889: This split has 179 training and 45 validation cases.
using pin_memory on device 0
using pin_memory on device 0
/opt/conda/envs/nnunet_test1/lib/python3.10/site-packages/torch/optim/lr_scheduler.py:28: UserWarning: The verbose parameter is deprecated. Please use get_last_lr() to access the learning rate.
  warnings.warn("The verbose parameter is deprecated. Please use get_last_lr() "

This is the configuration used by this training:
Configuration name: 3d_fullres
 {'data_identifier': 'nnUNetPlans_3d_fullres', 'preprocessor_name': 'DefaultPreprocessor', 'batch_size': 2, 'patch_size': [40, 224, 224], 'median_image_size_in_voxels': [95.0, 512.0, 512.0], 'spacing': [2.5, 0.8017575144767761, 0.8017575144767761], 'normalization_schemes': ['CTNormalization'], 'use_mask_for_norm': [False], 'resampling_fn_data': 'resample_data_or_seg_to_shape', 'resampling_fn_seg': 'resample_data_or_seg_to_shape', 'resampling_fn_data_kwargs': {'is_seg': False, 'order': 3, 'order_z': 0, 'force_separate_z': None}, 'resampling_fn_seg_kwargs': {'is_seg': True, 'order': 1, 'order_z': 0, 'force_separate_z': None}, 'resampling_fn_probabilities': 'resample_data_or_seg_to_shape', 'resampling_fn_probabilities_kwargs': {'is_seg': False, 'order': 1, 'order_z': 0, 'force_separate_z': None}, 'architecture': {'network_class_name': 'dynamic_network_architectures.architectures.unet.PlainConvUNet', 'arch_kwargs': {'n_stages': 6, 'features_per_stage': [32, 64, 128, 256, 320, 320], 'conv_op': 'torch.nn.modules.conv.Conv3d', 'kernel_sizes': [[1, 3, 3], [3, 3, 3], [3, 3, 3], [3, 3, 3], [3, 3, 3], [3, 3, 3]], 'strides': [[1, 1, 1], [1, 2, 2], [2, 2, 2], [2, 2, 2], [2, 2, 2], [1, 2, 2]], 'n_conv_per_stage': [2, 2, 2, 2, 2, 2], 'n_conv_per_stage_decoder': [2, 2, 2, 2, 2], 'conv_bias': True, 'norm_op': 'torch.nn.modules.instancenorm.InstanceNorm3d', 'norm_op_kwargs': {'eps': 1e-05, 'affine': True}, 'dropout_op': None, 'dropout_op_kwargs': None, 'nonlin': 'torch.nn.LeakyReLU', 'nonlin_kwargs': {'inplace': True}, 'deep_supervision': True}, '_kw_requires_import': ['conv_op', 'norm_op', 'dropout_op', 'nonlin']}, 'batch_dice': True} 

These are the global plan.json settings:
 {'dataset_name': 'Dataset007_Pancreas', 'plans_name': 'nnUNetPlans', 'original_median_spacing_after_transp': [2.5, 0.8017575144767761, 0.8017575144767761], 'original_median_shape_after_transp': [93, 512, 512], 'image_reader_writer': 'SimpleITKIO', 'transpose_forward': [0, 1, 2], 'transpose_backward': [0, 1, 2], 'experiment_planner_used': 'ExperimentPlanner', 'label_manager': 'LabelManager', 'foreground_intensity_properties_per_channel': {'0': {'max': 3071.0, 'mean': 81.05660247802734, 'median': 87.0, 'min': -932.0, 'percentile_00_5': -90.0, 'percentile_99_5': 214.0, 'std': 70.21927642822266}}} 

2024-06-04 00:42:41.793769: unpacking dataset...
2024-06-04 00:42:46.493493: unpacking done...
2024-06-04 00:42:46.527837: Unable to plot network architecture:
2024-06-04 00:42:46.527968: No module named 'hiddenlayer'
2024-06-04 00:42:46.537705: 
2024-06-04 00:42:46.537850: Epoch 0
2024-06-04 00:42:46.538045: Current learning rate: 0.01
Segmentation fault
```

Python=3.10.14
nnunetv2=2.4.2
torch=2.3.0+cu118

Thank you in advance.

**Comment by GregorKoehler** [comment]
Hi @jamesgwen,

can you provide some more info on the hardware you're using?

The "Segmentation fault" error you're encountering usually indicates a problem with accessing invalid memory areas, which can happen due to several reasons.
* GPU Memory Issues: Ensure your GPU has enough memory to handle the training with the given configuration
* PyTorch and CUDA Compatibility: Verify that the versions of PyTorch and CUDA are compatible.
* Memory Leaks or Fragmentation: Memory leaks or fragmentation can cause segmentation faults. Monitor the GPU memory usage using tools like `nvidia-smi` to ensure there are no unexpected memory spikes.
* Corrupted Data: Corrupted data files can sometimes lead to segmentation faults. Ensure that your dataset is not corrupted by verifying the integrity of the files.

**Comment by jamesgwen** [comment]
Hi @GregorKoehler 
Thanks for the quick reply.
Will look into the points you mention above.

Here is the output of nvidia-smi. 
```
Tue Jun  4 14:21:17 2024       
+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.86.10              Driver Version: 535.86.10    CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
| GPU  Name                 Persistence-M | Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp   Perf          Pwr:Usage/Cap |         Memory-Usage | GPU-Util  Compute M. |
|                                         |                      |               MIG M. |
|=========================================+======================+======================|
|   0  Tesla T4                       On  | 00000000:00:04.0 Off |                    0 |
| N/A   64C    P0              30W /  70W |   1358MiB / 15360MiB |      0%      Default |
|                                         |                      |                  N/A |
+-----------------------------------------+----------------------+----------------------+
                                                                                         
+---------------------------------------------------------------------------------------+
| Processes:                                                                            |
|  GPU   GI   CI        PID   Type   Process name                            GPU Memory |
|        ID   ID                                                             Usage      |
|=======================================================================================|
+---------------------------------------------------------------------------------------+
```

**Comment by GregorKoehler** [comment]
Did you verify the dataset integrity when running plan_and_preprocess? 
`nnUNetv2_plan_and_preprocess -d DATASET_ID --verify_dataset_integrity`
(following https://github.com/MIC-DKFZ/nnUNet/blob/master/documentation/how_to_use_nnunet.md)
Maybe some samples are corrupted?

**Comment by jamesgwen** [comment]
Yep. ```nnUNetv2_plan_and_preprocess -d DATASET_ID --verify_dataset_integrity``` ran well. No errors. Still getting the same behavior.

**Comment by mhson-kyle** [comment]
Hi! I am encountering the same error with you. 
I get Segmentation Fault error on 
nnunetv2=2.4.2
torch=2.3.1+cu118

Could not load library libcudnn_cnn_train.so.8. Error: /usr/local/cuda/lib64/libcudnn_cnn_train.so.8: undefined symbol: _ZN5cudnn3cnn34layerNormFwd_execute_internal_implERKNS_7backend11VariantPackEP11CUstream_stRNS0_18LayerNormFwdParamsERKNS1_20NormForwardOperationEmb, version libcudnn_cnn_infer.so.8 
error on 
nnunetv2=2.4.2
torch=2.3.1+cu121.

You mentioned that using torch=2.3.1+cu121 produced errors how did you resolve it.

> Hello, Giving an update. Was able to fix the error above with Pytorch 2.3.1+cu121. Tried to do more digging but couldn't figure out why it wasn't working with 2.2.0+cu121 or 2.3.0+cu118. Will close out the issue. Thank you for your advice/prompt responses.

**Comment by jamesgwen** [comment]
Hello,
I used this link to help debug. You have to change `LD_LIBRARY_PATH` to where pytorch is installed. 
https://discuss.pytorch.org/t/could-not-load-library-libcudnn-cnn-train-so-8-while-training-convnet/171334/7
